### PR TITLE
executor: fix issue #5132, update handle id with some other unique keys report Duplicate keys.

### LINKED ISSUE #5132: Update report unreasonable `Duplicate` error
TiDB:

```
mysql> CREATE TABLE `tt1` (
  `a` int(11) NOT NULL,
  `b` varchar(32) DEFAULT NULL,
  `c` varchar(32) DEFAULT NULL,
  PRIMARY KEY (`a`),
  UNIQUE KEY `b_idx` (`b`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into tt1 values(1, 'a', 'a');
Query OK, 1 row affected (0.00 sec)

mysql> insert into tt1 values(2, 'd', 'b');
Query OK, 1 row affected (0.00 sec)

mysql> select * from tt1;
+---+------+------+
| a | b    | c    |
+---+------+------+
| 1 | a    | a    |
| 2 | d    | b    |
+---+------+------+
2 rows in set (0.00 sec)

mysql> update tt1 set a=5 where c='b';
ERROR 1062 (23000): Duplicate entry 'd' for key 'b_idx'

mysql> update tt1 set a=5 where c='a';
ERROR 1062 (23000): Duplicate entry 'a' for key 'b_idx'
```

4. What version of TiDB are you using (`tidb-server -V`)?

```
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                               |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v1.1.0-alpha-106-g81652dbe
Git Commit Hash: 81652dbeefaa25a5116e19bc948c8ca268dd7a4d
Git Branch: master
UTC Build Time: 2017-11-17 04:46:48 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by winkyao**
fix #5132

### CODE REVIEW

**Comment by zimulala** [comment]
Please fix CI.

**Comment by winkyao** [comment]
@coocood @zimulala PTAL

File: executor/write.go [hunk 1]
```diff
@@ -40,12 +40,13 @@ var (
 // updateRecord updates the row specified by the handle `h`, from `oldData` to `newData`.
 // `modified` means which columns are really modified. It's used for secondary indices.
 // Length of `oldData` and `newData` equals to length of `t.WritableCols()`.
-func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, modified []bool, t table.Table, onDup bool) (bool, error) {
+func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, modified []bool, t table.Table, onDup, ignoreErr bool) (bool, error) {
```
Comment (shenli) [inline executor/write.go]: Add comment for the newly added argument.

File: table/tables/memory_tables.go [hunk 2]
```diff
@@ -186,7 +186,7 @@ func (t *MemoryTable) UpdateRecord(ctx context.Context, h int64, oldData, newDat
 }
 
 // AddRecord implements table.Table AddRecord interface.
-func (t *MemoryTable) AddRecord(ctx context.Context, r []types.Datum) (recordID int64, err error) {
+func (t *MemoryTable) AddRecord(ctx context.Context, r []types.Datum, recordDupChecked bool) (recordID int64, err error) {
```
Comment (coocood) [inline table/tables/memory_tables.go]: How about just use `skipCheck`, the name we are familiar with?

Comment (winkyao) [reply table/tables/memory_tables.go]: we already have skipCheck variable. it may be confused.

Comment (winkyao) [reply table/tables/memory_tables.go]: see line tables.go:419

Comment (coocood) [reply table/tables/memory_tables.go]: I know, but I think to use another name is more confusing.
It's the same thing in different scopes, like global variable and session variable.

**Comment by winkyao** [comment]
@coocood PTAL

File: table/table.go [hunk 3]
```diff
@@ -111,8 +111,9 @@ type Table interface {
 	// RecordKey returns the key in KV storage for the row.
 	RecordKey(h int64) kv.Key
 
-	// AddRecord inserts a row which should contain only public columns.
-	AddRecord(ctx context.Context, r []types.Datum) (recordID int64, err error)
+	// AddRecord inserts a row which should contain only public columns
+	// skipDupCheck indicate that recordID in r has been checked as not duplicate already.
```
Comment (coocood) [inline table/table.go]: Maybe `skipHandleCheck` is more accurate, it doesn't skip unique index check.

**Comment by winkyao** [comment]
@coocood PTAL

File: executor/write.go [hunk 4]
```diff
@@ -147,6 +159,22 @@ func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum,
 	return true, nil
 }
 
+func checkHandleExists(ctx context.Context, t table.Table, recordID int64) error {
+	txn := ctx.Txn()
+	// Check key exists.
+	recordKey := t.RecordKey(recordID)
+	e := kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key 'PRIMARY'", recordID)
+	txn.SetOption(kv.PresumeKeyNotExistsError, e)
+	defer txn.DelOption(kv.PresumeKeyNotExistsError)
```
Comment (coocood) [inline executor/write.go]: Above two lines can be removed.
It's only used for INSERT statement when `kv.PresumeKeyNotExists` option is set.

**Comment by coocood** [comment]
LGTM

**Comment by winkyao** [comment]
@shenli @zimulala PTAL

**Comment by shenli** [comment]
LGTM

File: table/tables/tables.go [hunk 5]
```diff
@@ -411,12 +411,12 @@ func (t *Table) genIndexKeyStr(colVals []types.Datum) (string, error) {
...
 
 // addIndices adds data into indices. If any key is duplicated, returns the original handle.
-func (t *Table) addIndices(ctx context.Context, recordID int64, r []types.Datum, bs *kv.BufferStore) (int64, error) {
+func (t *Table) addIndices(ctx context.Context, recordID int64, r []types.Datum, bs *kv.BufferStore, skipHandleCheck bool) (int64, error) {
 	txn := ctx.Txn()
 	// Clean up lazy check error environment
 	defer txn.DelOption(kv.PresumeKeyNotExistsError)
 	skipCheck := ctx.GetSessionVars().SkipConstraintCheck
-	if t.meta.PKIsHandle && !skipCheck {
+	if t.meta.PKIsHandle && !skipCheck && !skipHandleCheck {
```
Comment (zimulala) [inline table/tables/tables.go]: Can we use `checkHandleExists` directly?

Comment (zimulala) [reply table/tables/tables.go]: @winkyao

**Pull request merged**